[Android] Multiple place phones

strump opened the pull request:
POIs and places in OSM could have more than one phone number. Tag `phone` allows storing semicolon-separated list of numbers. E.g. https://osm.org/node/2143314850 https://osm.org/node/1693972693
But Organicmaps supports only single phone number field. This PR adds support of multiple phone numbers for a place.

## Place details

**Before:**

<img src="https://user-images.githubusercontent.com/720808/125298832-31013f00-e331-11eb-89f9-8c80eaf51899.png" width="250"/>

**After:**

<img src="https://user-images.githubusercontent.com/720808/125298857-38284d00-e331-11eb-9fd4-a71999872f15.png" width="250"/>

Each phone has separate line. Tap to call, long tap to copy.

## Call button

**Before:**

Pressing call button would initiate call to a concatenated phone number.

**After:**

If more than one phone is avaiilable a popup will appear

<img src="https://user-images.githubusercontent.com/720808/125299111-7de51580-e331-11eb-8bfa-e3c1b885b22e.png" width="250"/>

## Editing

**Before:**

Single edit text to edit all phone numbers

**After:**

New phone editing UI

https://user-images.githubusercontent.com/720808/125300069-5d698b00-e332-11eb-80b8-231c724a7222.mp4

## TODO

New string "Add phone" needs translation.

## Screenshots:

**Dark theme:**
<img src="https://user-images.githubusercontent.com/720808/126669452-06284327-9219-4de6-a5fb-127d1b49f363.png" width="500"/>

**RTL Layout:**
<img src="https://user-images.githubusercontent.com/720808/126669527-f79a0c07-ab46-4547-a7cb-81f234ad7d62.png" width="500"/>

Endem10n commented on android/res/layout/fragment_editor.xml:
```diff
@@ -144,9 +144,39 @@
+              android:layout_height="wrap_content"
+              android:layout_centerVertical="true"
+              android:layout_marginStart="@dimen/editor_margin_left"
+              android:layout_toLeftOf="@+id/edit_phone"
```
Redundant "android:layout_toLeftOf".  Min api level is 21.

strump replied:
@Endem10n I see that attribute `android:layout_toLeftOf` is used in fragment_editor.xml on Wi-Fi checkbox row.
Should I get replace RelativeLayout with LinearLayout to get rid of `android:layout_toLeftOf` property?
Or how to correctly layout phone icon, phone number text view, and edit phone buton in a single row?

Endem10n replied:
My point is that you already have in the next line "android:layout_toStartOf="@+id/edit_phone",  so all will work fine even without "android:layout_toLeftOf".

Endem10n commented on android/res/layout/fragment_editor.xml:
```diff
@@ -144,9 +144,39 @@
+          <TextView
+              android:id="@+id/edit_phone"
+              style="@style/MwmWidget.Editor.FieldLayout"
+              android:layout_marginLeft="12dp"
```
Redundant "android:layout_toLeftOf".  Min api level is 21.

Endem10n commented on android/res/layout/item_phone.xml:
```diff
@@ -0,0 +1,47 @@
+    android:orientation="horizontal"
+    android:paddingEnd="@dimen/margin_base"
+    android:paddingLeft="@dimen/margin_half_plus"
+    android:paddingRight="@dimen/margin_base"
```
Redundant tags "android:paddingLeft" and "android:paddingRight".  Min api level is 21.

Endem10n commented on android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:
```diff
@@ -1179,9 +1231,9 @@ public boolean onLongClick(View v)
-        break;
+      //case R.id.ll__place_phone:
+      //  items.add(mTvPhone.getText().toString());
+      //  break;
```
Possibly, should not comment, but just delete. It will stay in git history. Please, check in other places.

Endem10n commented on android/res/layout/item_phone.xml:
```diff
@@ -0,0 +1,47 @@
+        style="@style/MwmWidget.Editor.MetadataIcon"
+        tools:ignore="ContentDescription"
+        tools:src="@drawable/ic_phone"
+        android:layout_alignParentLeft="true"/>
```
Should use "android:layout_alignParentStart"

Endem10n commented on android/res/layout/item_phone.xml:
```diff
@@ -0,0 +1,47 @@
+        android:layout_centerVertical="true"
+        android:layout_marginLeft="@dimen/margin_half"
+        android:layout_toRightOf="@id/phone_icon"
+        android:layout_toLeftOf="@id/delete_icon"
```
Should change three lines to support RTL. Please, check in other places.

Endem10n commented on android/src/com/mapswithme/maps/editor/EditorHostFragment.java:
```diff
@@ -301,6 +311,13 @@ public void onClick(View v)
         break;
+      case PHONE:
+        final String phone = ((PhoneFragment) getChildFragmentManager().findFragmentByTag(PhoneFragment.class.getName())).getPhone();
+        if (Editor.nativeIsPhoneValid(phone)) {
```
Opening brace should be at the beginning of the next line according to project code style. Please, check in other places.

Endem10n commented on android/src/com/mapswithme/maps/editor/PhoneListAdapter.java:
```diff
@@ -0,0 +1,143 @@
+        }
+
+        public void setPhone(String phone) {
+            if(!mInput.getText().toString().equals(phone)) {
```
In case of single line "if" should not use braces according to project code style.

Endem10n commented on android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:
```diff
@@ -772,7 +798,8 @@ private void refreshDetails(@NonNull MapObject mapObject)
     String url = mapObject.getMetadata(Metadata.MetadataType.FMD_URL);
     refreshMetadataOrHide(TextUtils.isEmpty(website) ? url : website, mWebsite, mTvWebsite);
-    refreshMetadataOrHide(mapObject.getMetadata(Metadata.MetadataType.FMD_PHONE_NUMBER), mPhone, mTvPhone);
+    //refreshMetadataOrHide(mapObject.getMetadata(Metadata.MetadataType.FMD_PHONE_NUMBER), mPhone, mTvPhone);
```
Should not comment, but just delete.

Endem10n commented on android/res/layout/place_page_details.xml:
```diff
@@ -42,7 +42,7 @@
     <include layout="@layout/place_page_entrance"/>
 
-    <include layout="@layout/place_page_phone"/>
+    <include layout="@layout/place_page_phone_list"/>
```
nit:  As we already will have in this big vertical linear layout duplicating layout items, it would be great, if you will be able to refactor it to RecyclerView.

strump commented:
Added RecyclerView to display multiple phones on place details fragment.
Removed unused commented code. Fixed formatting.
Changed layout to support RTL languages and removed redundant attributes.

biodranik commented on android/res/values-nl/strings.xml:
```diff
@@ -598,6 +598,7 @@
 	<string name="bookmarks_error_title_list_name_already_taken">Deze naam is al in gebruik</string>
 	<string name="bookmarks_error_message_list_name_already_taken">Kies alstublieft een andere naam</string>
 	<string name="please_wait">Even geduld aub…</string>
+	<string name="phone_number">Telefoonnummer</string>
```
Did you manually add it? I see that these strings were already present in the master branch, in data/strings/strings.txt and in XML android resources.

strump replied:
Viacheslav Greshilov removed "phone_number" string in commit [#cca70039](https://github.com/organicmaps/organicmaps/commit/cca7003972cd540c6f1c7649250eb08fe5abc775) and generated strings.xml in commit [#2b46f73d](https://github.com/organicmaps/organicmaps/commit/2b46f73d3a8b65cfafe9383eb77bdf97a233f7a9)
I restored "phone_number" in strings.xml, but looks like I need to revert data/strings/strings.txt first

biodranik replied:
Wow, this string was unused, interesting. Thanks!

biodranik commented on android/src/com/mapswithme/maps/editor/EditorFragment.java:
```diff
@@ -170,14 +170,14 @@ public void onTextChanged(CharSequence s, int start, int before, int count)
 
     mPhone.setText(Editor.nativeGetPhone());
-    mPhone.addTextChangedListener(new StringUtils.SimpleTextWatcher()
+    /*mPhone.addTextChangedListener(new StringUtils.SimpleTextWatcher()
```
Why is it commented?

biodranik commented on android/src/com/mapswithme/maps/editor/EditorFragment.java:
```diff
@@ -222,7 +222,6 @@ boolean setEdits()
     Editor.nativeSetHouseNumber(mHouseNumber.getText().toString());
     Editor.nativeSetZipCode(mZipcode.getText().toString());
     Editor.nativeSetBuildingLevels(mBuildingLevels.getText().toString());
-    Editor.nativeSetPhone(mPhone.getText().toString());
```
Why is it removed? Are phones properly stored and sent to OSM?

strump replied:
I moved `Editor.nativeSetPhone(phone);` call to `EditorHostFragment` class simmilar to cuisine, opening house, street editors. Becase I created `PhoneFragment` to edit multiple phones, this fragment has "Save" button

biodranik commented on android/src/com/mapswithme/maps/editor/PhoneListAdapter.java:
```diff
@@ -0,0 +1,156 @@
+      final String p = phonesData.get(i).trim();
+      if (!TextUtils.isEmpty(p))
+      {
+        if (sb.length() > 0) sb.append(';');
```
We format simple ifs in two lines for easier debugging.

biodranik commented on android/src/com/mapswithme/maps/editor/PhoneListAdapter.java:
```diff
@@ -0,0 +1,156 @@
+
+      deleteButton = itemView.findViewById(R.id.delete_icon);
+      deleteButton.setOnClickListener(this);
+      //TODO: setting icons from code because icons defined in layout XML are white.
```
How to properly define icons from XML, so they will work in both light and dark themes?

biodranik replied:
nit: we insert spaces after slashes.

biodranik commented on android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:
```diff
@@ -537,9 +537,36 @@ private void onRouteRemoveBtnClicked()
+    }
+    else
+    {
+      //Show popup menu with all phones
```
Missed space after // and the dot at the end.

biodranik commented on android/src/com/mapswithme/maps/widget/placepage/PlacePhoneAdapter.java:
```diff
@@ -0,0 +1,122 @@
+    {
+      p = p.trim();
+      if (TextUtils.isEmpty(p)) continue;
+      Log.w(TAG, "adding phone: "+p);
```
We try not to pollute logs with debug strings (esp. warning ones).

strump replied:
I forgot to remove those logs. It's no use to log this text any more

biodranik reviewed (commented):
1. Did you test it for RTL languages?
2. Did you test dark and light themes?
3. Did you test it in a large tablet layout?

vng commented:
@Endem10n @biodranik Гляньте фиксы, я думаю пора мержить уже.

biodranik commented on data/strings/strings.txt:
```diff
@@ -14548,6 +14548,41 @@
+    tags = android
+    en = Phone number
+    ru = Номер телефона
+    ar = رقم الهاتف
```
@ma4ko What is the Bulgarian translation?

biodranik commented:
Uups, sorry, looks like conflicts should be resolved. We usually remove all commits with regenerated strings, rebase, and then generate strings again.

rtsisyk reviewed (approved):
Thank you for contribution! I don't see any blockers, but please try to address code style comments from other reviewers. This patch will be merged as soon as we fix localization in the master.

rtsisyk commented:
@strump, to save time, I squashed and rebased this PR.

See https://github.com/organicmaps/organicmaps/tree/multiple-place-phones

Please address comments from other reviewers and force-push the final version into this PR.

biodranik commented:
Let's polish it afterward to speed up and unblock other tasks with strings/localizations.
